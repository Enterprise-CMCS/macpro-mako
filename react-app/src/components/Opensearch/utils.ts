import { opensearch } from "shared-types";

const filterMapQueryReducer = (
  state: Record<opensearch.Filterable<any>["prefix"], any[]>,
  filter: opensearch.Filterable<any>,
) => {
  if (filter.type === "exists") {
    state[filter.prefix].push({
      exists: { field: filter.field },
    });
  }

  if (filter.value === undefined || filter.value == null) return state;

  // this was hoisted up since false is a valid "match" value
  if (filter.type === "match") {
    state[filter.prefix].push({
      match: { [filter.field]: filter.value },
    });
  }

  if (!filter.value) return state;

  if (filter.type === "terms") {
    state[filter.prefix].push({
      terms: { [filter.field]: filter.value },
    });
  }

  if (filter.type === "term") {
    state[filter.prefix].push({
      term: { [filter.field]: filter.value },
    });
  }

  if (filter.type === "range") {
    state[filter.prefix].push({
      range: { [filter.field]: filter.value },
    });
  }

  if (filter.type === "global_search") {
    state[filter.prefix].push({
      dis_max: {
        tie_breaker: 0.7,
        boost: 1.2,
        queries: ["id", "submitterName", "leadAnalystName"].map((FIELD) => ({
          wildcard: {
            [`${FIELD}.keyword`]: {
              value: `*${(filter.value as string).trim()}*`,
              case_insensitive: true,
            },
          },
        })),
      },
    });
  }

  return state;
};

export const filterQueryBuilder = (filters: opensearch.Filterable<any>[]) => {
  if (!filters?.length) return {};

  return {
    query: {
      bool: filters.reduce(filterMapQueryReducer, {
        must: [],
        must_not: [],
        should: [],
        filter: [],
      }),
    },
  };
};

export const paginationQueryBuilder = (pagination: opensearch.QueryState<any>["pagination"]) => {
  if (!pagination) return {};

  const from = (() => {
    if (
      !pagination?.number ||
      !pagination?.size ||
      pagination?.number < 1 ||
      pagination?.size < 1
    ) {
      return 0;
    }
    return pagination.number * pagination.size;
  })();

  return {
    size: pagination?.size && pagination?.size > 0 ? pagination?.size : 25,
    from,
  };
};

export const sortQueryBuilder = (sort: opensearch.QueryState<any>["sort"]) => {
  if (!sort?.field) return {};
  return { sort: [{ [sort.field]: sort.order || "asc" }] };
};

export const aggQueryBuilder = (aggs: opensearch.AggQuery<any>[]) => {
  if (!aggs?.length) return {};

  return {
    aggs: aggs.reduce((STATE, AGG) => {
      if (AGG?.name && AGG?.type && AGG?.field) {
        STATE[AGG.name] = {
          [AGG.type]: {
            field: AGG.field,
            order: { _term: "asc" },
            ...(AGG.size && { size: AGG.size }),
          },
        };
      }
      return STATE;
    }, {} as any),
  };
};

export const createSearchFilterable = (value?: string) => {
  if (!value) return [];
  return [
    {
      type: "global_search",
      field: "",
      value,
      prefix: "must",
    } as unknown as opensearch.Filterable<any>,
  ];
};

export const checkMultiFilter = (filters: opensearch.Filterable<any>[] = [], val: number = 0) => {
  return (
    filters.length >= val ||
    filters.some((filter) => Array.isArray(filter.value) && filter.value.length >= val)
  );
};
