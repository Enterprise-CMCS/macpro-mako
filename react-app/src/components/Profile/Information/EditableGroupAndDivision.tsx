import { zodResolver } from "@hookform/resolvers/zod";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import { API } from "aws-amplify";
import { LucidePencil } from "lucide-react";
import { useEffect, useMemo, useState } from "react";
import { useForm } from "react-hook-form";
import { z } from "zod";

import {
  Button,
  Dialog,
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
  Select,
  SelectContent,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/components";
import {
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
} from "@/components/Dialog";
import { divisionsType, groupDivision } from "@/features/sign-up/groupDivision";

const groupDivisionSchema = z.object({
  group: z.string().min(1, "A group selection is required."),
  division: z.string().min(1, "A division selection is required."),
});

type GroupAndDivisionFormProps = {
  group?: string;
  division?: string;
  onSubmit: (data: z.infer<typeof groupDivisionSchema>) => void;
  onCancel: () => void;
};

const GroupAndDivisionForm = ({
  group,
  division,
  onSubmit,
  onCancel,
}: GroupAndDivisionFormProps) => {
  const [isGroupSelectOpen, setIsGroupSelectOpen] = useState(false);
  const [isDivisionSelectOpen, setIsDivisionSelectOpen] = useState(false);
  const [availableDivisions, setAvailableDivisions] = useState<divisionsType[]>([]);

  const { initialGroupAbbr, initialDivisionAbbr } = useMemo(() => {
    const initialGroupObject = group ? groupDivision.find((g) => g.abbr === group) : undefined;
    return {
      initialGroupAbbr: initialGroupObject?.abbr || "",
      initialDivisionAbbr:
        initialGroupObject && division
          ? initialGroupObject.divisions.find((d) => d.abbr === division)?.abbr || ""
          : "",
    };
  }, [group, division]);

  const form = useForm<z.infer<typeof groupDivisionSchema>>({
    resolver: zodResolver(groupDivisionSchema),
    mode: "onChange",
    defaultValues: {
      group: initialGroupAbbr,
      division: initialDivisionAbbr,
    },
  });

  const selectedGroup = form.watch("group");
  const selectedDivision = form.watch("division");

  useEffect(() => {
    const groupData = groupDivision.find((g) => g.abbr === selectedGroup);
    setAvailableDivisions(groupData?.divisions ?? []);
  }, [selectedGroup]);

  const handleEscapeKeyDown = (event: KeyboardEvent) => {
    if ((isGroupSelectOpen || isDivisionSelectOpen) && event.key === "Escape") {
      event.preventDefault();
    }
  };

  const isButtonDisabled = !selectedGroup || !selectedDivision;

  return (
    <DialogContent onEscapeKeyDown={handleEscapeKeyDown}>
      <DialogHeader>
        <DialogTitle>Select your Group and Division</DialogTitle>
      </DialogHeader>
      <DialogDescription className="sr-only">
        This form will collect the necessary details to edit your group and division. Please fill
        out all required fields.
      </DialogDescription>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8">
          <FormField
            control={form.control}
            name="group"
            render={({ field }) => (
              <FormItem>
                <FormLabel className="font-semibold">Group</FormLabel>
                <Select
                  onValueChange={(value) => {
                    field.onChange(value);
                    form.setValue("division", "");
                  }}
                  value={field.value}
                  open={isGroupSelectOpen}
                  onOpenChange={setIsGroupSelectOpen}
                >
                  <FormControl>
                    <SelectTrigger>
                      <SelectValue placeholder="Select a group" />
                    </SelectTrigger>
                  </FormControl>
                  <SelectContent>
                    {groupDivision.map((group) => (
                      <SelectItem key={group.id} value={group.abbr}>
                        <div className="flex items-start gap-x-2">
                          <strong className="w-16 flex-shrink-0 text-left">{group.abbr}</strong>
                          <span>{group.name}</span>
                        </div>
                      </SelectItem>
                    ))}
                  </SelectContent>
                </Select>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="division"
            render={({ field }) => (
              <FormItem>
                <FormLabel className="font-semibold">Division</FormLabel>
                <Select
                  onValueChange={field.onChange}
                  value={field.value}
                  open={isDivisionSelectOpen}
                  onOpenChange={setIsDivisionSelectOpen}
                  disabled={!selectedGroup || availableDivisions.length === 0}
                >
                  <FormControl>
                    <SelectTrigger>
                      <SelectValue placeholder="Select a division" />
                    </SelectTrigger>
                  </FormControl>
                  <SelectContent>
                    {availableDivisions.map((division) => (
                      <SelectItem key={division.id} value={division.abbr || division.name}>
                        {division.abbr ? (
                          <div className="flex items-start gap-x-2">
                            <strong className="w-16 flex-shrink-0 text-left">
                              {division.abbr}
                            </strong>
                            <span>{division.name}</span>
                          </div>
                        ) : (
                          division.name
                        )}
                      </SelectItem>
                    ))}
                  </SelectContent>
                </Select>
                <FormMessage />
              </FormItem>
            )}
          />
          <DialogFooter className="flex sm:justify-start">
            <Button type="submit" disabled={isButtonDisabled || form.formState.isSubmitting}>
              {form.formState.isSubmitting ? "Submitting..." : "Submit"}
            </Button>
            <Button type="button" variant="link" onClick={onCancel}>
              Cancel
            </Button>
          </DialogFooter>
        </form>
      </Form>
    </DialogContent>
  );
};

type EditableGroupAndDivisionProps = {
  group?: string;
  division?: string;
  allowEdits?: boolean;
  email: string;
};

export const EditableGroupAndDivision = ({
  group,
  division,
  allowEdits,
  email,
}: EditableGroupAndDivisionProps) => {
  const [isDialogOpen, setIsDialogOpen] = useState(false);
  const [formKey, setFormKey] = useState(0);
  const queryClient = useQueryClient();

  const { mutate: submitGroupDivision } = useMutation({
    mutationFn: (data: z.infer<typeof groupDivisionSchema>) => {
      return API.post("os", "/submitGroupDivision", {
        body: {
          userEmail: email,
          ...data,
        },
      });
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["userDetails"] });
      setIsDialogOpen(false);
    },
  });

  const handleOpenDialog = () => {
    setFormKey((prevKey) => prevKey + 1);
    setIsDialogOpen(true);
  };

  const handleSubmit = (formData: z.infer<typeof groupDivisionSchema>) => {
    submitGroupDivision(formData);
  };

  const handleCancel = () => {
    setIsDialogOpen(false);
  };

  return (
    <div className="leading-9">
      <h3 className="font-bold flex items-center">
        Group & Division{" "}
        {allowEdits && (
          <Button
            type="button"
            variant="ghost"
            size="icon"
            onClick={handleOpenDialog}
            aria-label="Edit Group and Division"
          >
            <LucidePencil className="w-5" />
          </Button>
        )}
      </h3>

      <p>
        {group || "N/A"}/{division || "N/A"}
      </p>

      <Dialog open={isDialogOpen} onOpenChange={setIsDialogOpen}>
        <GroupAndDivisionForm
          key={formKey}
          group={group}
          division={division}
          onSubmit={handleSubmit}
          onCancel={handleCancel}
        />
      </Dialog>
    </div>
  );
};
