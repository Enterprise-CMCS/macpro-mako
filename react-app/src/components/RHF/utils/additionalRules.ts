import { RegisterOptions } from "react-hook-form";
import { RuleGenerator, SortFuncs, AdditionalRule } from "shared-types/forms";

export const sortFunctions: {
  [x in SortFuncs]: (a: string, b: string) => number;
} = {
  noSort: () => 0,
  reverseSort: (a, b) => b.localeCompare(a),
};

// New function to sort options from lowest to highest
export function sortOptionsLowestToHighest<
  T extends { value: string | number },
>(options: T[]): T[] {
  return [...options].sort((a, b) => {
    const aValue = typeof a.value === "string" ? parseFloat(a.value) : a.value;
    const bValue = typeof b.value === "string" ? parseFloat(b.value) : b.value;
    return aValue - bValue;
  });
}

export function stringCompare(
  a: { label: string },
  b: { label: string },
): number {
  const aIsNumber = !isNaN(parseFloat(a.label));
  const bIsNumber = !isNaN(parseFloat(b.label));

  if (aIsNumber && bIsNumber) {
    return parseFloat(a.label) - parseFloat(b.label);
  } else {
    return a.label.localeCompare(b.label);
  }
}

export const ruleGenerator: RuleGenerator = (rules, addtnlRules) => {
  if (!rules && !addtnlRules) return undefined;
  const simpleRules = rules ?? {};
  const customRules = addtnlRules
    ? { validate: addtnlRules.reduce(valReducer, {}) }
    : {};

  return { ...simpleRules, ...customRules };
};

export const valReducer = (
  valSet: RegisterOptions["validate"],
  rule: AdditionalRule,
  index: number,
): RegisterOptions["validate"] => {
  const valName = `${rule.type}_${index}`;

  switch (rule.type) {
    case "lessThanField":
      return {
        ...valSet,
        [valName]: (value, fields) => {
          if (
            !rule.strictGreater &&
            parseFloat(value) <= parseFloat(fields[rule.fieldName])
          )
            return true;
          else if (parseFloat(value) < parseFloat(fields[rule.fieldName]))
            return true;
          return rule.message;
        },
      };
    case "greaterThanField":
      return {
        ...valSet,
        [valName]: (value, fields) => {
          if (
            !rule.strictGreater &&
            parseFloat(value) >= parseFloat(fields[rule.fieldName])
          )
            return true;
          else if (parseFloat(value) > parseFloat(fields[rule.fieldName]))
            return true;
          return rule.message;
        },
      };
    case "cannotCoexist":
      return {
        ...valSet,
        [valName]: (value, fields) => {
          if (value !== undefined && fields[rule.fieldName] === undefined)
            return true;
          if (value === undefined && fields[rule.fieldName] !== undefined)
            return true;
          return rule.message;
        },
      };
    case "noGapsOrOrverlaps":
      return {
        ...valSet,
        [valName]: (_, fields) => {
          const fieldArray = fields[rule.fieldName];
          if (
            !fieldArray ||
            !Array.isArray(fieldArray) ||
            fieldArray.length <= 1
          ) {
            return true; // No validation needed for 0 or 1 entry
          }

          const fromField = rule.fromField;
          const toField = rule.toField;

          const range = fieldArray.map((item: any) => ({
            from: parseInt(item[fromField], 10),
            to: parseInt(item[toField], 10),
          }));

          // Sort ranges by from value
          range.sort((a, b) => a.from - b.from);

          // Check for overlaps and gaps
          for (let i = 1; i < range.length; i++) {
            if (range[i].from <= range[i - 1].to) {
              return "No age overlaps allowed";
            }
            if (range[i].from > range[i - 1].to + 1) {
              return "No gaps between ages allowed";
            }
          }

          return true; // No issues found
        },
      };
    case "toGreaterThanFrom":
      return {
        ...valSet,
        [valName]: (value, formValues) => {
          const fromValue = parseInt(value, 10);
          const fieldArray = formValues[rule.fieldName];

          if (!fieldArray || !Array.isArray(fieldArray)) {
            return true; // Skip validation if field array is not valid
          }

          const currentIndex = fieldArray.findIndex(
            (item) => item[rule.fromField] === value,
          );
          if (currentIndex === -1) {
            return true; // Skip validation if current item is not found
          }

          const toValue = parseInt(fieldArray[currentIndex][rule.toField], 10);

          if (isNaN(fromValue) || isNaN(toValue)) {
            return true; // Skip validation if values are not valid numbers
          }

          return (
            toValue > fromValue ||
            rule.message ||
            "To age must be greater than From age"
          );
        },
      };
    default:
      return { ...valSet };
  }
};
