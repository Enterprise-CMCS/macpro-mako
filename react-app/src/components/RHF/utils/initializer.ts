import {
  RHFOption,
  RHFSlotProps,
  FormGroup,
  FormSchema,
} from "shared-types/forms";

type GL = Record<string, unknown>;

export const formGroupInitializer =
  (parentId?: string) => (ACC: GL, FORM: FormGroup) => {
    FORM.slots.reduce(slotInitializer(parentId), ACC);
    return ACC;
  };

export const slotInitializer =
  (parentId?: string) =>
  (ACC: GL, SLOT: RHFSlotProps): GL => {
    const adjustedName = `${parentId ?? ""}${SLOT.name}`;

    const optionReducer = (OPT: RHFOption) => {
      if (OPT.form) OPT.form.reduce(formGroupInitializer(parentId), ACC);
      if (OPT.slots) OPT.slots.reduce(slotInitializer(parentId), ACC);
      return ACC;
    };

    const fieldInitializer = (ACC1: GL, SLOTC: RHFSlotProps): GL => {
      if (SLOTC.rhf === "FieldArray") {
        return {
          ...ACC1,
          [SLOTC.name]: [SLOTC.fields?.reduce(fieldInitializer, {})],
        };
      }

      return { ...ACC1, ...slotInitializer()(ACC1, SLOTC) };
    };

    switch (SLOT.rhf) {
      case "TextDisplay":
        break;
      case "Switch":
        ACC[adjustedName] = false;
        break;
      case "Checkbox":
        SLOT.props?.options.forEach(optionReducer);
        ACC[adjustedName] = [];
        break;
      case "Radio":
        SLOT.props?.options.forEach(optionReducer);
        ACC[adjustedName] = "";
        break;
      case "FieldArray":
        ACC[adjustedName] = [SLOT.fields?.reduce(fieldInitializer, {})];
        break;
      case "WrappedGroup":
        ACC = { ...ACC, ...SLOT.fields?.reduce(fieldInitializer, {}) };
        break;
      case "Upload":
        ACC[adjustedName] = [];
        break;
      // If switching from undefined to a Date causes an error from RHF,
      // DatePickerProps in inputs.ts may need to be adjusted:
      case "DatePicker":
        ACC[adjustedName] = undefined;
        break;
      case "Input":
      case "Select":
      case "Multiselect":
      case "Textarea":
      default:
        ACC[adjustedName] = "";
        break;
    }

    return ACC;
  };

export const documentInitializer = (document: FormSchema) => {
  return document.sections.reduce((ACC, SEC) => {
    SEC.form.reduce(
      formGroupInitializer(`${document.formId}_${SEC.sectionId}_`),
      ACC,
    );
    return ACC;
  }, {});
};
