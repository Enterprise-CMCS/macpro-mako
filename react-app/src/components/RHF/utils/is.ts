import { ValidationRule } from "react-hook-form";

export const INPUT_VALIDATION_RULES = {
  max: "max",
  min: "min",
  maxLength: "maxLength",
  minLength: "minLength",
  pattern: "pattern",
  required: "required",
  validate: "validate",
} as const;
export type InputValidationRules = typeof INPUT_VALIDATION_RULES;
export type ERROR = Record<string, string>;
export type MaxType = InputValidationRules["max"] | InputValidationRules["maxLength"];

export type MinType = InputValidationRules["min"] | InputValidationRules["minLength"];

export const isFunction = (value: unknown): value is typeof Function => typeof value === "function";

export const isNullOrUndefined = (value: unknown): value is null | undefined => value == null;

export const isUndefined = (val: unknown): val is undefined => val === undefined;

export const isDateObject = (value: unknown): value is Date => value instanceof Date;

export const isObjectType = (value: unknown) => typeof value === "object";

export const isString = (value: unknown): value is string => typeof value === "string";

export const isObject = <T extends object>(value: unknown): value is T =>
  !isNullOrUndefined(value) && !Array.isArray(value) && isObjectType(value) && !isDateObject(value);

export const isRegex = (value: unknown): value is RegExp => value instanceof RegExp;

export const getValueAndMessage = (validationData?: ValidationRule) =>
  isObject(validationData) && !isRegex(validationData)
    ? validationData
    : {
        value: validationData,
        message: "",
      };
