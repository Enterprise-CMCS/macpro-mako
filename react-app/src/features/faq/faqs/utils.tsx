import { cn } from "@/utils";

export interface Template {
  title: string;
  text?: string;
  href: string;
  subtext?: string[];
}

export const helpDeskContact = {
  email: "OneMAC_Helpdesk@cms.hhs.gov",
  phone: "(833) 228-2540",
};

export const slugify = (text: string) =>
  text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/(^-|-$)/g, "")
    .substring(0, 40);

export const handleSupportLinkClick = (type: string) => (e: React.MouseEvent<HTMLElement>) => {
  const text = e.currentTarget.textContent?.trim() || "unknown";
  window.gtag("event", `support_click_${type}_${slugify(text)}`, {
    event_category: "Support",
    event_label: text,
  });
};

export const PdfLink = ({
  href,
  title,
  text,
  label,
  className = "underline hover:no-underline",
}: {
  href: string;
  title: string;
  text?: string;
  label: string;
  className?: string;
}) => (
  <a
    className={cn("text-primary", className)}
    href={href}
    download
    rel="noopener noreferrer"
    target="_blank"
    onClick={handleSupportLinkClick(label)}
  >
    {title}
    {text && `: ${text}`}
  </a>
);

export const PdfList = ({
  list,
  label,
  ulClassName = "list-disc pl-6 space-y-2",
}: {
  list: Template[];
  label: string;
  ulClassName?: string;
}) => (
  <ul className={ulClassName} role="list">
    {list.map((pdf) => (
      <li key={pdf.title}>
        <PdfLink href={pdf.href} title={pdf.title} text={pdf.text} label={label} className="" />
        {pdf.subtext && (
          <ul className="list-disc pl-7 space-y-1" role="list">
            {pdf.subtext.map((sub, index) => (
              <li key={index} className="text-sm text-gray-600">
                {sub}
              </li>
            ))}
          </ul>
        )}
      </li>
    ))}
  </ul>
);
