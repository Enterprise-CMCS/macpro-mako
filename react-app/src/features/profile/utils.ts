import { UserRole } from "shared-types/events/legacy-user";

import { StateAccess } from "@/api";
import { convertStateAbbrToFullName } from "@/utils";

export const userRoleMap = {
  defaultcmsuser: "CMS Read-only User",
  cmsroleapprover: "CMS Role Approver",
  cmsreviewer: "CMS Read-only User",
  statesystemadmin: "State System Admin",
  helpdesk: "Helpdesk",
  statesubmitter: "State Submitter",
  systemadmin: "System Admin",
};

export const stateAccessRoles: UserRole[] = [
  "statesubmitter",
  "statesystemadmin",
  "cmsroleapprover",
  "systemadmin",
];

export const orderStateAccess = (accesses: StateAccess[]) => {
  if (!accesses || !accesses.length) return;
  // sort revoked states seprately and add to
  const activeStates = accesses.filter((x: StateAccess) => x.status != "revoked");
  const revokedStates = accesses.filter((x: StateAccess) => x.status == "revoked");

  const compare = (a: StateAccess, b: StateAccess) => {
    const stateA = convertStateAbbrToFullName(a.territory);
    const stateB = convertStateAbbrToFullName(b.territory);

    if (stateA < stateB) return -1;
    if (stateA > stateB) return 1;
    return 0;
  };

  const sorted = activeStates.sort(compare).concat(revokedStates.sort(compare));

  return sorted;
};

// if user has no active roles, show pending state(s)
// show state(s) for latest active role
export const filterStateAccess = (userDetails, userProfile) => {
  if (!userProfile?.stateAccess || userProfile.stateAccess.length < 1) return [];
  return userDetails?.role
    ? userProfile.stateAccess.filter(
        (access: StateAccess) => access.role === userDetails.role && access.territory !== "ZZ",
      )
    : userProfile.stateAccess.filter((access: StateAccess) => access.territory !== "ZZ");
};
