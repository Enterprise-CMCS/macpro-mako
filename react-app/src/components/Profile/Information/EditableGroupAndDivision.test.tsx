import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { afterEach, describe, expect, it, vi } from "vitest";

import { renderWithQueryClient } from "@/utils/test-helpers";

import { EditableGroupAndDivision } from "./EditableGroupAndDivision";

const mockMutate = vi.fn();

afterEach(() => {
  document.querySelectorAll("[data-scroll-lock]").forEach((el) => el.remove());
  document.body.style.overflow = "auto";
  document.body.style.pointerEvents = "auto";
  document.querySelectorAll("[data-radix-portal]").forEach((el) => el.remove());
});

describe("EditableGroupAndDivision", () => {
  const user = userEvent.setup();

  it("should open the dialog when the edit button is clicked", async () => {
    renderWithQueryClient(
      <EditableGroupAndDivision
        group="Group A"
        division="Division 1"
        allowEdits
        email="test@example.com"
      />,
    );

    const editButton = screen.getByRole("button", { name: /edit group and division/i });
    await user.click(editButton);

    expect(screen.getByRole("dialog")).toBeInTheDocument();
  });

  it("should have a disabled submit button until both fields are selected", async () => {
    renderWithQueryClient(<EditableGroupAndDivision allowEdits email="test@example.com" />);

    const editButton = screen.getByRole("button", { name: /edit group and division/i });
    await user.click(editButton);

    const submitButton = screen.getByRole("button", { name: /submit/i });
    expect(submitButton).toBeDisabled();

    await user.click(screen.getByTestId("group-select"));
    await user.click(screen.getByText("DMCO"));

    expect(submitButton).toBeDisabled();
  });

  it("should call the mutation and close the dialog on successful submission", async () => {
    renderWithQueryClient(<EditableGroupAndDivision allowEdits email="test@example.com" />);

    const editButton = screen.getByRole("button", { name: /edit group and division/i });
    await user.click(editButton);

    await user.click(screen.getByTestId("group-select"));
    await user.click(screen.getByText("DMCO"));

    await user.click(screen.getByTestId("division-select"));
    await user.click(screen.getByText("Division of Managed Care Operations"));

    const submitButton = screen.getByRole("button", { name: /submit/i });
    await user.click(submitButton);

    await waitFor(() => {
      expect(mockMutate).toHaveBeenCalledWith({
        group: "DMCO",
        division: "DMCO",
      });
    });
  });

  it("should close the dialog when the cancel button is clicked", async () => {
    renderWithQueryClient(<EditableGroupAndDivision allowEdits email="test@example.com" />);

    const editButton = screen.getByRole("button", { name: /edit group and division/i });
    await user.click(editButton);

    const cancelButton = screen.getByRole("button", { name: /cancel/i });
    await user.click(cancelButton);

    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
  });
});
