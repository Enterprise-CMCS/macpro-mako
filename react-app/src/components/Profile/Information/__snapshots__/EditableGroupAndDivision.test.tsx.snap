// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditableGroupAndDivision  > should render correctly with allowEdits false 1`] = `
<DocumentFragment>
  <div
    class="leading-9"
  >
    <h3
      class="font-bold flex items-center"
    >
      Group & Division 
    </h3>
    <p>
      -- -- -- --
    </p>
  </div>
  ,
</DocumentFragment>
`;

exports[`EditableGroupAndDivision  > should render correctly with allowEdits true 1`] = `
<DocumentFragment>
  <div
    class="leading-9"
  >
    <h3
      class="font-bold flex items-center"
    >
      Group & Division 
      <button
        aria-label="Edit Group and Division"
        class="inline-flex items-center justify-center rounded-md text-[16px] font-bold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none hover:bg-accent hover:text-accent-foreground disabled:opacity-50 h-10 w-10"
        type="button"
      >
        <svg
          class="lucide lucide-pencil w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
          />
          <path
            d="m15 5 4 4"
          />
        </svg>
      </button>
    </h3>
    <p>
      -- -- -- --
    </p>
  </div>
  ,
</DocumentFragment>
`;

exports[`EditableGroupAndDivision  > should render correctly with group and division 1`] = `
<DocumentFragment>
  <div
    class="leading-9"
  >
    <h3
      class="font-bold flex items-center"
    >
      Group & Division 
      <button
        aria-label="Edit Group and Division"
        class="inline-flex items-center justify-center rounded-md text-[16px] font-bold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none hover:bg-accent hover:text-accent-foreground disabled:opacity-50 h-10 w-10"
        type="button"
      >
        <svg
          class="lucide lucide-pencil w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
          />
          <path
            d="m15 5 4 4"
          />
        </svg>
      </button>
    </h3>
    <p>
      Group A/Division 1
    </p>
  </div>
  ,
</DocumentFragment>
`;

exports[`EditableGroupAndDivision  > should render correctly without group and division 1`] = `
<DocumentFragment>
  <div
    class="leading-9"
  >
    <h3
      class="font-bold flex items-center"
    >
      Group & Division 
      <button
        aria-label="Edit Group and Division"
        class="inline-flex items-center justify-center rounded-md text-[16px] font-bold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none hover:bg-accent hover:text-accent-foreground disabled:opacity-50 h-10 w-10"
        type="button"
      >
        <svg
          class="lucide lucide-pencil w-5"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
          />
          <path
            d="m15 5 4 4"
          />
        </svg>
      </button>
    </h3>
    <p>
      -- -- -- --
    </p>
  </div>
  ,
</DocumentFragment>
`;
